src/masternodes/loan.cpp

```diff
@@ -241,66 +249,119 @@ static base_uint<128> ToHigherPrecision(CAmount amount, uint32_t height)
     return amountHP;
 }
 
-CAmount TotalInterest(const CInterestRateV2& rate, uint32_t height)
+CNegativeInterest TotalInterestCalculation(const CInterestRateV3& rate, const uint32_t height, const CAmount loanAmount, const CAmount tokenInterest, const CAmount schemeInterest)
+{
+    CNegativeInterest interest;
+    if (rate.interestPerBlock.negative) {
+
+        base_uint<128> reduction{};
+        auto reductionRate{rate.interestPerBlock};
+        auto interestToHeight{rate.interestToHeight};
+        auto loanAmountHP{ToHigherPrecision(loanAmount, height)};
+
+        for (size_t i{}; i < height - rate.height; ++i) {
+            auto reductionStep = reductionRate.amount;
+            if (interestToHeight > 0) {
+                if (interestToHeight > reductionStep) {
+                    interestToHeight -= reductionStep;
+                    reduction += reductionStep;
+                    reductionStep = 0;
+                } else {
+                    interestToHeight = 0;
+                    reductionStep -= interestToHeight;
+                    reduction += interestToHeight;
+                }
+            }
+            const auto reductionCeil = reductionStep;
+            if (loanAmountHP > reductionCeil) {
+                loanAmountHP -= reductionCeil;
+                reduction += reductionCeil;
+            } else {
+                // Loan amount reduced to zero
+                reduction += loanAmountHP;
+                break;
+            }
+            reductionRate = InterestPerBlockCalculationV3(CeilInterest(loanAmountHP, std::numeric_limits<uint32_t>::max()), tokenInterest, schemeInterest);
+        }
```

Comment:
I have to ask again: why doing this complicated calculation which makes everything harder (calculations, explaining it to users) and acts as the worst case for users?

positive interest is on purpose (for performance reasions) not compounding:
loan of 1 token, interestRatePerBlock = 0.1 (based on this 1 token loan) we get 
block 0 : loan 1 , interest 0
block 1: loan 1, interest 0.1
block 2: loan 1, interest 0.2
... easy to explain, easy to calculate

negative interest should (IMHO) be the same logic. interestPerBlock = -0.1
block 0: loan 1, interest 0
block 1: loan 1, interest -0.1
block 2: loan 1, interest -0.2
....
at block 10 we see loan + interest <= 0 and we auto close the loan. But until then we accumulate the negative interest based on the loan. 

The way it currently is implemented, there is a different interest effect if you opened a 1 token loan 1 month ago and accumulated positive interest and a fresh loan of 1 token:
with existing interest on the loan you profit, because the negative rate is calculated from the loan, not loan + interest, but the negative interest first reduces your existing interest (so for longer time you get higher negative interestPerBlock).
while for the fresh loan, the negative interest reduces your loan value and therefore the negativeInterestPerBlock.

This will be super hard to explain, specially since its bad for the user and now different to how it was done for positive interest.

Reply:
This will make life a lot easier, not just to explain to the users but on the implementation front. Spoke to the team about this and will start updating now.